### PR TITLE
Ability to extend default function result handling

### LINKED ISSUE #757: Spring Cloud Function adapter for Azure (class org.springframework.cloud.function.adapter.azure.FunctionInvoker) does not “transmit” reactive behavior.
We want to write reactive java function, with Azure Function.
But now the returned value of handler is not of type Mono or Flux.
The handler is implemented in handleRequest() method of class org.springframework.cloud.function.adapter.azure.FunctionInvoker.

A piece of code of this class is shown bellow.
We can see that there is blocking call.

```java
   if (output instanceof Publisher) {
            if (FunctionTypeUtils.isMono(function.getOutputType())) {
                return (O) this.convertOutputIfNecessary(input, Mono.from((Publisher) output).blockOptional().get());
            }
            else {
                List resultList = new ArrayList<>();
                for (Object resultItem : Flux.from((Publisher) output).toIterable()) {
                    if (resultItem instanceof Collection) {
                        resultList.addAll((Collection) resultItem);
                    }
                    else {
                        if (Collection.class.isAssignableFrom(FunctionTypeUtils.getRawType(function.getInputType()))
                                && !Collection.class.isAssignableFrom(FunctionTypeUtils.getRawType(function.getOutputType()))) {
                            return (O) this.convertOutputIfNecessary(input, resultItem);
                        }
                        else {
                            resultList.add(resultItem);
                        }
                    }
                }
                return (O) this.convertOutputIfNecessary(input, resultList);
            }
        }
```

Thanks.

### PR DESCRIPTION
**Pull request by onobc**
This gives custom function invokers the ability to post-process the function result and optionally return something different than the default behavior. 

#### TODO
- [x] Add docs on how to run the sample

Fixes #757

### CODE REVIEW

**Comment by onobc** [comment]
@alain-marcel
@RouxAntoine

See if this would give you what you need to "get at" the Mono/Flux and do something different w/ it.

File: spring-cloud-function-adapters/spring-cloud-function-adapter-azure/src/main/java/org/springframework/cloud/function/adapter/azure/FunctionInvoker.java [hunk 1]
```diff
@@ -119,37 +120,117 @@ else if (function == null && StringUtils.hasText(functionDefinition) && APPLICAT
...
-		Object output = function.apply(enhancedInput);
-		if (output instanceof Publisher) {
-			if (FunctionTypeUtils.isMono(function.getOutputType())) {
-				return (O) this.convertOutputIfNecessary(input, Mono.from((Publisher) output).blockOptional().get());
+		Object functionResult = function.apply(enhancedInput);
+
+		if (!(functionResult instanceof Publisher)) {
+			return postProcessImperativeFunctionResult(input, enhancedInput, functionResult, function, executionContext);
+		}
+		return postProcessReactiveFunctionResult(input, enhancedInput, (Publisher<?>) functionResult, function, executionContext);
```
Comment (RouxAntoine) [inline spring-cloud-function-adapters/spring-cloud-function-adapter-azure/src/main/java/org/springframework/cloud/function/adapter/azure/FunctionInvoker.java:134]: I think the opposed condition is simpler to understand

```suggestion
		if (functionResult instanceof Publisher) {
			return postProcessReactiveFunctionResult(input, enhancedInput, (Publisher<?>) functionResult, function, executionContext);
		}
		return postProcessImperativeFunctionResult(input, enhancedInput, functionResult, function, executionContext);
```

**Review by RouxAntoine — COMMENTED** [review]
I like the idea to provide `postProcessXXXResullt` hook method. great 👌

**Comment by RouxAntoine** [comment]
LGTM

**Pull request closed without merging**